### ISSUE TITLE
brave-vertx module

### ISSUE THREAD

**Issue by adriancole** [openzipkin/brave#307]
Requests to trace vertx have come up more often in gitter https://gitter.im/openzipkin/zipkin

There is an implementation here https://github.com/emmanuelidi/vertx-zipkin, and we've also gotten requests for other versions of vertx

Does it make sense to make a multi-module repo openzipkin/brave-vertx ? If ok with @emmanuelidi, we could actually move his repo there and reorganize it to support multiple versions and the publication process we have. Alternatively, we could start again, yet another option is to just push requests to emmanuelidi/vertx-zipkin

Thoughts? @mkargar @lintuxlin @emmanuelidi @fzakaria @hosuaby @cescoffier @vietj

**Comment by fzakaria** [comment]
@adriancole great to hear from you.
I think vert.x addition to brave would be useful, at the very a least an example that is non-blocking would be a boon to similar frameworks.
I looked originally at the repo you provided but found it not a useful implementation of Zipkin on the Vertx framework.

Ultimately the shitty part about non-blocking frameworks is the inability to leverage threadlocal to pass to arbitrary depths the traceIds.
I got something working at my current workplace, but it equates to passing down a context map everywhere -- have yet to check in that branch.

perhaps there is some better integration I'm missing. I know we do something with SFL4J MDC in our codebase. We make sure all non-blocking code goes through a specific Util class so the MDC gets transferred & reset between threads.

tl;dr; restart from scratch and provide any non-blocking framework (even barebones netty) a zipkin implementation.

**Comment by cescoffier** [comment]
As said by @fzakaria, you can't rely on thread locals in Vert.x, and also the number of jumps between _contexts_ can be very high (even co-localized components interacts using asynchronous messages).

The current approach is to pass some kind of tracer object (or correlation id), however, it can be very cumbersome, and you may want to hide it. For this reason, we are working on providing a way to build your own "locals" system that tracks context jumps and let you propagate data between the context. This should be available in Vert.x 3.4.0 (ETA January-February)

**Comment by mkargar** [comment]
The implementation above is a good start for server side but doesn't cover the client side of the story (e.g. HttpClientRequest, HttpClientResponse)

**Comment by adriancole** [comment]
for those curious about other async libraries, the one with most recent maintenance is CXF #304

Also, the propagation story is a well-known concern. The new api #180 (which takes quite an effort to ensure weaves in, starting with #300), will treat explicit or user-supplied span activation as a first class concept. This will be notably easier as you don't have to differentiate between a client, server or local span like you do now. Essentially, you have an active span (propagated however makes sense) or you don't.

**Comment by vietj** [comment]
@adriancole hi,  @emmanuelidi is supposed to contribute his Vert.x Core changes to the codebase that allow to integrate ZipKin as he did in his current prototype.

**Comment by emmanuelidi** [comment]
I'm fine moving the code wherever you guys would like and can help with any coding you think is necessary. The current implementation still needs some work around documentation and I think I have some local code clean-up changes. I would like to finish that to make it easier for others to evaluate and consume. 

fyi: the changes I'm proposing for `vertx-core` are in my fork[1]. I had been waiting on completing the above prior to opening a PR to flush out any problems.

[1] https://github.com/eclipse/vert.x/compare/master...emmanuelidi:master+tracing

**Comment by vietj** [comment]
@adriancole on vertx-dev list there is an on-going discussion about the changes necessary in Vert.x Core for enabling Zipkin support and @emmanuelidi contribution : https://groups.google.com/d/msg/vertx/9rHzX3vVg8Y/3aDGhd_ZEgAJ so I've good hope it's gonna move forward soon!

**Comment by andreas-schilling** [comment]
Are there any news on this?

**Comment by adriancole** [comment]
I've heard nothing, but I'd recommend to whoever is doing this to use the recent apis maybe looking at ratpack as an example https://github.com/hyleung/ratpack-zipkin

**Comment by adriancole** [comment]
Brave 4.3 is being synced now, I highly recommend re-starting this discussion

**Comment by fzakaria** [comment]
We're the changes merged into vert.x master yet ?
Last I checked they weren't.

**Comment by adriancole** [comment]
I think it is safe to say we shouldn't block on https://github.com/eclipse/vert.x/pull/1787 work as it is stalled out, maybe for good reasons, ye we've continue to get requests, yet again today from @allienna

A more pragmatic way could be to port http handling code from @emmanuelidi's original work on Brave 3 to the new Brave 4 apis https://github.com/emmanuelidi/vertx-zipkin/tree/master/src/main/java/io/vertx/ext/zipkin/web

Seems that's how Red Hat approached it in OpenTracing (@pavolloffay @objectiser @cescoffier @pmlopes) https://github.com/opentracing-contrib/java-vertx-web/pull/2 I think we shouldn't let best be the enemy of the good.

We can always revisit if/when propagation work carries forward, but meanwhile stop the "how about now" loop we've been in for a year plus, or the choice to require the Brave 3 to use vert.x (Brave 3 won't get any more updates even if you can wire it with Brave 4 using TracerAdapter)

@emmanuelidi many thanks for putting the initial work in, I'll ping you on a PR and will note in the README any limitations our integration tests found.

**Comment by vietj** [comment]
@adriancole the main issue is that eclipse/vert.x#1787 won't be enough to provide a reliable implementation and we need (in vertx) to come up with a more universal solution to this problem

**Comment by adriancole** [comment]
Ok so maybe linking to that as the end goal isn't a good idea.

Still I wonder if getting tests that fail due to lack of a more universal
solution isnt worth it.. we do this sometimes in other instrumentation,
adding a test that ignores or asserts fail due to inability to continue a
context implicitly from an arbitrary callback.

If we can in code show what doesnt work, we have something to come back to
later.. unless you feel like nothing can work and it is pointless. (I dont
know as havent tried it).

If some limited functionality can work, then is it mostly a matter of
linking to a more appropriate issue for comprehensive solution?

**Comment by adriancole** [comment]
it was easier than I thought to get all of our tests passing. PTAL https://github.com/openzipkin/brave/pull/597 (doesn't have http client yet)
